dnf/transaction_sr.py

```diff
@@ -0,0 +1,508 @@
+
+        # Override configuration options that affect how the transaction is resolved
+        # TODO if used through API, we shouldn't change the global state here?
+        self._base.conf.clean_requirements_on_remove = False
+        self._base.conf.install_weak_deps = False
+
+        fn = self._filename
+        errors = []
+
+        for pkg_data in self._rpms:
+            try:
+                self._replay_pkg_action(pkg_data)
+            except TransactionError as e:
+                errors.append(e)
+
+        for group_data in self._groups:
+            try:
+                action = group_data["action"]
+                group_id = group_data["id"]
+
+                if action == "Install":
+                    try:
+                        pkg_types = libdnf.transaction.stringToCompsPackageType(group_data["package_types"])
+                    except libdnf.error.Error as e:
+                        errors.append(TransactionError(str(e)))
+                        continue
+
+                    self._base.group_install(group_id, pkg_types)
+                elif action == "Upgrade":
+                    self._base.group_upgrade(group_id)
+                elif action == "Removed":
+                    self._base.group_remove(group_id)
+                else:
+                    errors.append(TransactionError(
+                        _('Unexpected value of group action "{action}" for group "{group}".')
+                            .format(action=action, group=group_id)
+                    ))
+            except KeyError as e:
+                errors.append(TransactionError(
+                    _('Missing object key "{key}" in a group.').format(key=e.args[0])
+                ))
+            except ValueError as e:
+                # group does not exist
+                errors.append(TransactionError(str(e)))
+            except dnf.exceptions.CompsError as e:
+                # TODO need to distinguish the two cases below. The first one
+                # is --ignore-installed and the second one is --skip-missing:
+                #
+                # - installed group not found for upgrade/remove
+                # - group not available anymore for upgrade
+                #
+                # Should I create a new exception inheriting from CompsError to
+                # maintain compatibility and distinguish the two cases?
+                errors.append(TransactionError(str(e)))
+
+        for env_data in self._environments:
+            try:
+                action = env_data["action"]
+                env_id = env_data["id"]
+
+                if action == "Install":
+                    try:
+                        pkg_types = libdnf.transaction.stringToCompsPackageType(env_data["package_types"])
+                    except libdnf.error.Error as e:
+                        errors.append(TransactionError(str(e)))
+                        continue
+
+                    self._base.environment_install(env_id, pkg_types)
```

Comment:
I don't understand what you're suggesting, can you elaborate?

Reply:
Well try to replay `dnf group install <foo> -x <foo-pkg>`. (<foo-pkg> is a package name that is part of the group but not required by other members of the group). It will result that transaction will be not replayable because during replay `foo-pkg` appears in the transaction.
Or try to replay `dnf group install core -x '*'`

Reply:
Ok, yes, I know of that limitation. We've actually discussed this with @dmach long time ago and sort-of agreed to solve this later, as this is dragging out for way too long.

What I don't understand and wanted you to elaborate on was your proposed solution to this.

Reply:
You can take an inspiration from dnf code:
```
    def _group_install(self, group_id, pkg_types, exclude=None, strict=True, exclude_groups=None):
        assert dnf.util.is_string_type(group_id)
        comps_group = self.comps._group_by_id(group_id)
        if not comps_group:
            raise ValueError(_("Group id '%s' does not exist.") % ucd(group_id))

        swdb_group = self.history.group.new(group_id, comps_group.name, comps_group.ui_name, pkg_types)
        for i in comps_group.packages_iter():
            swdb_group.addPackage(i.name, False, i.type)
        self.history.group.install(swdb_group)
```

Reply:
I'm sorry, but I have no idea what you mean. Could you please provide a reasonably thorough description of your proposal, your hint is vague and just leaves me guessing.

Reply:
I suggest to modify only swdb according to stored data in json and not to also try to install rpms according to comp group. Is that better?

Reply:
We should probably discuss that directly, seems to me this has a number of issues (like not actually installing the packages means the transaction is not fully replayed...?)